File: eventbrite.js
const puppeteer = require('puppeteer');
const fs = require('fs');

// Location-specific configurations
const LOCATION_CONFIGS = {
    'santarosa': {
        'url': 'https://www.eventbrite.com/d/ca--santa-rosa/all-events/',
        'timezone': 'US/Pacific',
        'prodid': '-//Eventbrite Santa Rosa Events//',
        'calname': 'Eventbrite Santa Rosa Events'
    },
    'bloomington': {
        'url': 'https://www.eventbrite.com/d/in--bloomington/all-events/',
        'timezone': 'America/Indiana/Indianapolis',
        'prodid': '-//Eventbrite Bloomington Events//',
        'calname': 'Eventbrite Bloomington Events'
    }
};

// Parse command line arguments
const args = process.argv.slice(2);
if (args.length !== 3) {
    console.error('Usage: node common_eventbrite.js <location> <year> <month>');
    process.exit(1);
}

const location = args[0].toLowerCase();
const targetYear = parseInt(args[1], 10);
const targetMonth = parseInt(args[2], 10);

if (!LOCATION_CONFIGS[location] || isNaN(targetYear) || isNaN(targetMonth) || targetMonth < 1 || targetMonth > 12) {
    console.error('Invalid location, year, or month. Please provide a valid location, year, and month (1-12).');
    process.exit(1);
}

const locationConfig = LOCATION_CONFIGS[location];

async function scrapeEvents(url, targetYear, targetMonth) {
    const browser = await puppeteer.launch({ 
        headless: 'new',
        args: [
            '--no-sandbox', 
            '--disable-setuid-sandbox',
            '--disable-dev-shm-usage',
            '--disable-gpu',
            '--window-size=1920,1080'
        ]
    });
    const page = await browser.newPage();
    page.setDefaultNavigationTimeout(60000);
    
    // Set a realistic user agent to avoid bot detection
    await page.setUserAgent('Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36');
    await page.setViewport({ width: 1920, height: 1080 });

    console.log(`Navigating to ${url}...`);
    await page.goto(url, { waitUntil: 'networkidle2', timeout: 60000 });
    
    // Wait for events to load
    try {
        await page.waitForSelector('[data-testid="search-event"]', { timeout: 30000 });
        console.log('Events loaded successfully');
    } catch (e) {
        console.error('Failed to find events on page. Page might be blocked or have different structure.');
        const html = await page.content();
        console.log('Page content preview:', html.substring(0, 500));
        await browser.close();
        return [];
    }

    let eventsInMonth = [];
    const maxPages = 40; // Maximum number of pages to process

    async function extractEventsOnPage() {
        console.log("Extracting events data...");
        const events = await page.evaluate((targetYear, targetMonth) => {
            const eventElements = document.querySelectorAll('[data-testid="search-event"]');
            const eventList = [];
            
            // Helper to get today/tomorrow dates
            const now = new Date();
            const today = new Date(now.getFullYear(), now.getMonth(), now.getDate());
            const tomorrow = new Date(today);
            tomorrow.setDate(tomorrow.getDate() + 1);

            eventElements.forEach(event => {
                const name = event.querySelector('h3')?.innerText || 'No title provided';

                const dateTimeElement = Array.from(event.querySelectorAll('p')).find(p => /\b(?:AM|PM)\b/.test(p.innerText));

                let formattedDateTime = 'Invalid date';
                if (dateTimeElement) {
                    const dateTimeText = dateTimeElement.innerText.trim();
                    try {
                        // Handle "Today • 8:00 PM" or "Tomorrow • 7:00 PM"
                        const todayMatch = dateTimeText.match(/Today\s*•\s*(\d{1,2}:\d{2}\s*[APM]+)/i);
                        const tomorrowMatch = dateTimeText.match(/Tomorrow\s*•\s*(\d{1,2}:\d{2}\s*[APM]+)/i);
                        
                        // Handle "Sat, Mar 7 • 1:30 PM"
                        const dateMatch = dateTimeText.match(/(\w+),?\s*(\w+\s+\d{1,2})\s*•?\s*(\d{1,2}:\d{2}\s*[APM]+)/);
                        
                        let dateTime = null;
                        
                        if (todayMatch) {
                            const time = todayMatch[1];
                            const [hours, minutes] = time.match(/(\d+):(\d+)/).slice(1);
                            const isPM = /PM/i.test(time);
                            let hour = parseInt(hours);
                            if (isPM && hour !== 12) hour += 12;
                            if (!isPM && hour === 12) hour = 0;
                            dateTime = new Date(today);
                            dateTime.setHours(hour, parseInt(minutes), 0, 0);
                        } else if (tomorrowMatch) {
                            const time = tomorrowMatch[1];
                            const [hours, minutes] = time.match(/(\d+):(\d+)/).slice(1);
                            const isPM = /PM/i.test(time);
                            let hour = parseInt(hours);
                            if (isPM && hour !== 12) hour += 12;
                            if (!isPM && hour === 12) hour = 0;
                            dateTime = new Date(tomorrow);
                            dateTime.setHours(hour, parseInt(minutes), 0, 0);
                        } else if (dateMatch) {
                            const [_, dayOfWeek, date, time] = dateMatch;
                            // Try current year first, then next year if date is in the past
                            let fullDate = `${date}, ${targetYear} ${time}`;
                            dateTime = new Date(fullDate);
                            // If the date is more than 2 months in the past, assume next year
                            if (dateTime < new Date(now.getTime() - 60*24*60*60*1000)) {
                                fullDate = `${date}, ${targetYear + 1} ${time}`;
                                dateTime = new Date(fullDate);
                            }
                        }
                        
                        if (dateTime && !isNaN(dateTime.getTime())) {
                            formattedDateTime = dateTime.toISOString().replace(/[-:]/g, '').split('.')[0] + 'Z';
                        }
                    } catch (error) {
                        // Silent fail, will be marked as Invalid date
                    }
                }

                const location = event.querySelector('[data-event-location]')?.getAttribute('data-event-location') || 'No location provided';
                const link = event.querySelector('a.event-card-link')?.href || 'No link provided';

                eventList.push({
                    name,
                    date: formattedDateTime,
                    location,
                    link
                });
            });

            return eventList;
        }, targetYear, targetMonth);

        eventsInMonth.push(events);
    }

    async function handlePagination() {
        let pageNumber = 1;

        while (pageNumber <= maxPages) {
            await extractEventsOnPage();

            console.log(`\nAccumulated events after page ${pageNumber}:`);
            eventsInMonth.forEach((pageEvents, index) => {
                console.log(`${index + 1}`);
                pageEvents.forEach(event => {
                    console.log(`  ${event.date}`);
                });
            });

            const lastPageEvents = eventsInMonth[eventsInMonth.length - 1];
            
            // Count valid events and check if any are in target month
            let validEventCount = 0;
            let eventsInTargetMonth = 0;
            let eventsPastTargetMonth = 0;
            
            for (let event of lastPageEvents) {
                if (event.date === 'Invalid date') continue;
                validEventCount++;
                
                const year = parseInt(event.date.slice(0, 4), 10);
                const month = parseInt(event.date.slice(4, 6), 10);

                if (year === targetYear && month === targetMonth) {
                    eventsInTargetMonth++;
                } else if (year > targetYear || (year === targetYear && month > targetMonth)) {
                    eventsPastTargetMonth++;
                }
            }
            
            console.log(`Page ${pageNumber}: ${lastPageEvents.length} events, ${validEventCount} valid dates, ${eventsInTargetMonth} in target month, ${eventsPastTargetMonth} past target`);

            // Only stop if we have valid events and ALL of them are past the target month
            if (validEventCount > 0 && eventsInTargetMonth === 0 && eventsPastTargetMonth === validEventCount) {
                console.log("\nExit condition met. All valid events on this page are past the target month. Stopping pagination.");
                break;
            }
            
            // Also stop if we got no events at all (empty page)
            if (lastPageEvents.length === 0) {
                console.log("\nNo events found on page. Stopping pagination.");
                break;
            }

            if (pageNumber >= maxPages) {
                console.log("\nMaximum page limit reached. Stopping pagination.");
                break;
            }

            console.log("Checking for 'Next Page' button...");
            const nextPageButton = await page.$('button[data-testid="page-next"]');
            if (!nextPageButton) {
                console.log("No 'Next Page' button found. Ending pagination.");
                break;
            }

            console.log("Clicking 'Next Page' button...");
            await nextPageButton.click();

            console.log("Waiting for the next page of events to load...");
            try {
                await page.waitForSelector('[data-testid="search-event"]', { visible: true, timeout: 60000 });
                console.log("Next page loaded, continuing to extract events.");
            } catch (error) {
                console.error("Error waiting for the next page to load:", error);
                break;
            }

            pageNumber++;
        }
    }

    await handlePagination();
    await browser.close();

    return eventsInMonth.flat();
}

function createICSContent(events, locationConfig, targetYear, targetMonth) {
    let icsContent = `BEGIN:VCALENDAR
VERSION:2.0
CALSCALE:GREGORIAN
METHOD:PUBLISH
PRODID:${locationConfig.prodid}
X-WR-CALNAME:${locationConfig.calname}
X-WR-TIMEZONE:${locationConfig.timezone}
BEGIN:VTIMEZONE
TZID:${locationConfig.timezone}
X-LIC-LOCATION:${locationConfig.timezone}
BEGIN:STANDARD
TZOFFSETFROM:-0700
TZOFFSETTO:-0800
TZNAME:PST
DTSTART:19701101T020000
RRULE:FREQ=YEARLY;BYMONTH=11;BYDAY=1SU
END:STANDARD
BEGIN:DAYLIGHT
TZOFFSETFROM:-0800
TZOFFSETTO:-0700
TZNAME:PDT
DTSTART:19700308T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=2SU
END:DAYLIGHT
END:VTIMEZONE
`;

    events.forEach(event => {
        if (event.date !== 'Invalid date') {
            const year = parseInt(event.date.slice(0, 4), 10);
            const month = parseInt(event.date.slice(4, 6), 10);

            if (year === targetYear && month === targetMonth) {
                icsContent += `BEGIN:VEVENT
SUMMARY:${event.name}
DTSTART:${event.date}
LOCATION:${event.location}
URL:${event.link}
DESCRIPTION:${event.name}
END:VEVENT
`;
            } else if (year > targetYear || (year === targetYear && month > targetMonth)) {
                console.log(`Excluding event "${event.name}" as it occurs after the target month.`);
            }
        } else {
            console.warn(`Skipping event "${event.name}" due to invalid date format.`);
        }
    });

    icsContent += "END:VCALENDAR";
    return icsContent;
}

async function main() {
    console.log(`Scraping Eventbrite events for ${location} in ${targetYear}-${targetMonth}`);
    const events = await scrapeEvents(locationConfig.url, targetYear, targetMonth);
    const icsContent = createICSContent(events, locationConfig, targetYear, targetMonth);

    const paddedTargetMonth = targetMonth.toString().padStart(2, '0');       
    const fileName = `./${location}/eventbrite_${targetYear}_${paddedTargetMonth}.ics`;
    fs.writeFileSync(fileName, icsContent);
    console.log(`${fileName} file generated successfully!`);
}

main().catch(console.error);

File: eventbrite.py
#!/usr/bin/env python3
"""
Eventbrite scraper using JSON-LD structured data.
No browser needed - just HTTP requests.
"""

import json
import re
import sys
from datetime import datetime
from icalendar import Calendar, Event, vText
import requests

LOCATION_CONFIGS = {
    'santarosa': {
        'url': 'https://www.eventbrite.com/d/ca--santa-rosa/all-events/',
        'timezone': 'US/Pacific',
        'prodid': '-//Eventbrite Santa Rosa Events//',
        'calname': 'Eventbrite Santa Rosa Events'
    },
    'bloomington': {
        'url': 'https://www.eventbrite.com/d/in--bloomington/all-events/',
        'timezone': 'America/Indiana/Indianapolis',
        'prodid': '-//Eventbrite Bloomington Events//',
        'calname': 'Eventbrite Bloomington Events'
    }
}

def fetch_events_from_page(url, session):
    """Fetch and parse JSON-LD events from a single page."""
    resp = session.get(url, timeout=30)
    resp.raise_for_status()
    
    matches = re.findall(r'<script type="application/ld\+json">(.*?)</script>', resp.text, re.DOTALL)
    
    events = []
    for m in matches:
        try:
            data = json.loads(m)
            if isinstance(data, dict) and data.get('@type') == 'ItemList':
                for item in data.get('itemListElement', []):
                    event = item.get('item', {})
                    if event.get('@type') == 'Event':
                        events.append(event)
        except json.JSONDecodeError:
            continue
    
    return events

def scrape_eventbrite(location, target_year, target_month, max_pages=20):
    """Scrape Eventbrite events for a location and month."""
    config = LOCATION_CONFIGS.get(location)
    if not config:
        raise ValueError(f"Unknown location: {location}")
    
    headers = {
        'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.9',
        'Accept-Encoding': 'gzip, deflate, br',
        'Cache-Control': 'no-cache',
        'Pragma': 'no-cache',
        'Sec-Ch-Ua': '"Not_A Brand";v="8", "Chromium";v="120", "Google Chrome";v="120"',
        'Sec-Ch-Ua-Mobile': '?0',
        'Sec-Ch-Ua-Platform': '"Linux"',
        'Sec-Fetch-Dest': 'document',
        'Sec-Fetch-Mode': 'navigate',
        'Sec-Fetch-Site': 'none',
        'Sec-Fetch-User': '?1',
        'Upgrade-Insecure-Requests': '1',
    }
    
    all_events = []
    seen_urls = set()
    
    # Use a session to maintain cookies across requests
    session = requests.Session()
    session.headers.update(headers)
    
    for page in range(1, max_pages + 1):
        if page == 1:
            url = config['url']
        else:
            url = f"{config['url']}?page={page}"
        
        print(f"Fetching page {page}: {url}")
        
        try:
            events = fetch_events_from_page(url, session)
        except Exception as e:
            print(f"Error fetching page {page}: {e}")
            break
        
        if not events:
            print(f"No events found on page {page}, stopping.")
            break
        
        # Track if we've gone past target month
        events_in_month = 0
        events_past_month = 0
        
        for event in events:
            event_url = event.get('url', '')
            if event_url in seen_urls:
                continue
            seen_urls.add(event_url)
            
            start_date = event.get('startDate', '')
            if not start_date:
                continue
            
            try:
                # Parse date (format: 2026-02-21 or 2026-02-21T19:00:00)
                if 'T' in start_date:
                    dt = datetime.fromisoformat(start_date.replace('Z', '+00:00'))
                else:
                    dt = datetime.strptime(start_date, '%Y-%m-%d')
                
                if dt.year == target_year and dt.month == target_month:
                    events_in_month += 1
                    all_events.append(event)
                elif dt.year > target_year or (dt.year == target_year and dt.month > target_month):
                    events_past_month += 1
            except ValueError as e:
                print(f"Could not parse date '{start_date}': {e}")
                continue
        
        print(f"  Found {len(events)} events, {events_in_month} in target month, {events_past_month} past target")
        
        # Stop if all events on page are past target month
        if events_in_month == 0 and events_past_month > 0:
            print("All events past target month, stopping pagination.")
            break
    
    return all_events, config

def create_ics(events, config, target_year, target_month):
    """Create ICS calendar from events."""
    cal = Calendar()
    cal.add('prodid', config['prodid'])
    cal.add('version', '2.0')
    cal.add('calscale', 'GREGORIAN')
    cal.add('method', 'PUBLISH')
    cal.add('x-wr-calname', config['calname'])
    cal.add('x-wr-timezone', config['timezone'])
    
    for event_data in events:
        event = Event()
        
        event.add('summary', event_data.get('name', 'Untitled Event'))
        
        start_date = event_data.get('startDate', '')
        end_date = event_data.get('endDate', start_date)
        
        try:
            if 'T' in start_date:
                dt_start = datetime.fromisoformat(start_date.replace('Z', '+00:00'))
            else:
                dt_start = datetime.strptime(start_date, '%Y-%m-%d')
            event.add('dtstart', dt_start)
            
            if end_date:
                if 'T' in end_date:
                    dt_end = datetime.fromisoformat(end_date.replace('Z', '+00:00'))
                else:
                    dt_end = datetime.strptime(end_date, '%Y-%m-%d')
                event.add('dtend', dt_end)
        except ValueError:
            continue
        
        # Location
        location = event_data.get('location', {})
        if isinstance(location, dict):
            loc_name = location.get('name', '')
            address = location.get('address', {})
            if isinstance(address, dict):
                street = address.get('streetAddress', '')
                city = address.get('addressLocality', '')
                state = address.get('addressRegion', '')
                loc_str = ', '.join(filter(None, [loc_name, street, city, state]))
            else:
                loc_str = loc_name
            if loc_str:
                event.add('location', vText(loc_str))
        
        # URL
        if event_data.get('url'):
            event.add('url', event_data['url'])
        
        # Description
        if event_data.get('description'):
            event.add('description', event_data['description'])
        
        # UID
        event.add('uid', f"{event_data.get('url', '')}-{start_date}")
        
        cal.add_component(event)
    
    return cal

def main():
    if len(sys.argv) != 4:
        print("Usage: python eventbrite_jsonld.py <location> <year> <month>")
        print("  Locations: santarosa, bloomington")
        sys.exit(1)
    
    location = sys.argv[1].lower()
    target_year = int(sys.argv[2])
    target_month = int(sys.argv[3])
    
    if location not in LOCATION_CONFIGS:
        print(f"Unknown location: {location}")
        print(f"Available: {', '.join(LOCATION_CONFIGS.keys())}")
        sys.exit(1)
    
    print(f"Scraping Eventbrite for {location}, {target_year}-{target_month:02d}")
    
    events, config = scrape_eventbrite(location, target_year, target_month)
    
    print(f"\nTotal events in {target_year}-{target_month:02d}: {len(events)}")
    
    cal = create_ics(events, config, target_year, target_month)
    
    filename = f"./{location}/eventbrite_{target_year}_{target_month:02d}.ics"
    with open(filename, 'wb') as f:
        f.write(cal.to_ical())
    
    print(f"Wrote {filename}")

if __name__ == '__main__':
    main()
